leader-election-impl/src/test/java/com/palantir/leader/proxy/AwaitingLeadershipProxyTest.java

```diff
@@ -333,6 +339,148 @@ public void shouldClearDelegateIfLeadershipTokenIsRefreshedByAnotherProxy() thro
+                .when(mock)
+                .val();
+        Supplier<MyCloseable> factory = createSpyableSupplier(mock);
+        LeadershipCoordinator coordinator = LeadershipCoordinator.create(leaderElectionService);
+        MyCloseable proxy = AwaitingLeadershipProxy.newProxyInstance(MyCloseable.class, factory, coordinator);
+
+        MyCloseable bystander = mock(MyCloseable.class);
+        Supplier<MyCloseable> bystanderFactory = createSpyableSupplier(bystander);
+        MyCloseable bystanderProxy =
+                AwaitingLeadershipProxy.newProxyInstance(MyCloseable.class, bystanderFactory, coordinator);
+
+        waitForLeadershipToBeGained();
+
+        bystanderProxy.val();
+
+        assertThatLoggableExceptionThrownBy(proxy::val)
+                .isInstanceOf(SuspectedNotCurrentLeaderException.class)
+                .hasMessage("There is one imposter among us");
+
+        assertThatCode(proxy::val)
+                .as("the underlying mock can be called again, if we actually maintained leadership")
+                .doesNotThrowAnyException();
+
+        assertThatCode(() -> {
+                    proxy.val();
+                    bystanderProxy.val();
+                    proxy.val();
+                    bystanderProxy.val();
+                    proxy.val();
+                })
+                .as("the proxies are still able to forward requests")
+                .doesNotThrowAnyException();
+
+        verify(factory, times(2)).get();
+        verify(bystanderFactory).get();
+        verify(mock).close();
+        verify(bystander, never()).close();
+    }
+
+    @Test
+    public void shouldLoseLeadershipOnSuspectedNotCurrentLeaderWhenLeadershipGenuinelyLost()
+            throws InterruptedException, IOException {
+        MyCloseable mock = mock(MyCloseable.class);
+        MyCloseable proxy = AwaitingLeadershipProxy.newProxyInstance(
+                MyCloseable.class, () -> mock, LeadershipCoordinator.create(leaderElectionService));
+        waitForLeadershipToBeGained();
+
+        doAnswer(_invocation -> {
+                    loseLeadershipOnLeaderElectionService(StillLeadingStatus.NOT_LEADING);
+                    throw new SuspectedNotCurrentLeaderException("There is one imposter among us");
+                })
+                .when(mock)
+                .val();
+
+        assertThatLoggableExceptionThrownBy(proxy::val)
+                .isInstanceOf(NotCurrentLeaderException.class)
+                .hasMessageContaining("method invoked on a non-leader (leadership lost)");
+
+        // Another call is required before we clear existing resources.
+        assertThatLoggableExceptionThrownBy(proxy::val)
+                .isInstanceOf(NotCurrentLeaderException.class)
+                .hasMessageContaining("method invoked on a non-leader");
+        verify(mock).close();
+    }
+
+    @Test
+    public void shouldLoseLeadershipOnSuspectedNotCurrentLeaderWhenNoQuorum() throws InterruptedException, IOException {
```

Comment:
Any reason you chose not just to parameterise the test with an EnumSource + Exclude Leading?

Reply:
This module is not on JUnit 5. I've added a TODO though.